lib/awful/widget/layoutlist.lua

```diff
@@ -0,0 +1,442 @@
+        self._private.layout:set_spacing(
+            self._private.style.spacing or beautiful.tasklist_spacing or 0
+        )
+    end
+
+    assert(self._private.layout.is_widget)
+
+    update(self)
+
+    self:emit_signal("widget::layout_changed")
+    self:emit_signal("widget::redraw_needed")
+end
+
+function layoutlist:set_widget_template(widget_template)
+    self._private.widget_template = widget_template
+
+    -- Remove the existing instances
+    self._private.data = {}
+
+    -- Prevent a race condition when the constructor loop to initialize the
+    -- arguments.
+    if self._private.layout then
+        update(self)
+    end
+
+    self:emit_signal("widget::layout_changed")
+    self:emit_signal("widget::redraw_needed")
+end
+
+function layoutlist:layout(_, width, height)
+    if self._private.layout then
+        return { wibox.widget.base.place_widget_at(self._private.layout, 0, 0, width, height) }
+    end
+end
+
+function layoutlist:fit(context, width, height)
+    if not self._private.layout then
+        return 0, 0
+    end
+
+    return wibox.widget.base.fit_widget(self, context, self._private.layout, width, height)
+end
+
+--- Create a layout list.
+--
+-- @tparam table args
+-- @tparam widget args.layout The widget layout (not to be confused with client
+--  layout).
+-- @tparam table args.buttons A table with buttons binding to set.
+-- @tparam[opt=awful.widget.layoutlist.source.for_screen] function args.source A
+--  function to generate the list of layouts.
+-- @tparam[opt] table args.widget_template A custom widget to be used for each action.
+-- @tparam[opt=ascreen.focused()] screen args.screen A screen
+-- @tparam[opt=nil] table args.buttons The list of `awful.buttons`.
+-- @tparam[opt={}] table args.style Extra look and feel parameters
+-- @tparam boolean args.style.disable_icon
+-- @tparam boolean args.style.disable_name
+-- @tparam string|pattern args.style.fg_normal
+-- @tparam string|pattern args.style.bg_normal
+-- @tparam string|pattern args.style.fg_selected
+-- @tparam string|pattern args.style.bg_selected
+-- @tparam string args.style.font
+-- @tparam string args.style.font_selected
+-- @tparam string args.style.align *left*, *right* or *center*
+-- @tparam number args.style.spacing
+-- @tparam gears.shape args.style.shape
+-- @tparam number args.style.shape_border_width
+-- @tparam string|pattern args.style.shape_border_color
+-- @tparam gears.shape args.style.shape_selected
+-- @tparam string|pattern args.style.shape_border_width_selected
+-- @tparam string|pattern args.style.shape_border_color_selected
+-- @treturn widget The action widget.
+-- @function naughty.widget.layoutlist
+
+local is_connected, instances = false, setmetatable({}, {__mode = "v"})
+
+local function update_common()
+    for _, ll in ipairs(instances) do
+        reload_cache(ll)
+        update(ll)
+    end
+end
+
+local function new(_, args)
+    local wdg = wibox.widget.base.make_widget(nil, nil, {
```

Comment:
here the widget is called **wdg** in some other widget is call **ret** it's would be good to pick a single name to be consistent across the widget library and I vote for an [meaningful](https://medium.com/coding-skills/clean-code-101-meaningful-names-and-functions-bf450456d90c) name :) like **new_widget**